A basic representation of Holonomic Functions and Differential Operator

shubhamtibra opened the pull request:
- [x] Added HoloFunc class
- [x]  Added DifferentialOperator class
- [x] Added Tests

Basic classes to represent a holonomic function and differential operator. Addition of Holonomic Function is implemented.
Ping @certik @fredrik-johansson

shubhamtibra committed:
A preliminary class for Differential Operator and Holonomic Functions.

kshitij10496 commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,133 @@
+
+from __future__ import print_function, division
+from sympy import symbols, poly, diff, Symbol, sstr, pretty_print, printing, Function, dsolve
+
```
Leave 2 blank lines here

kshitij10496 commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,133 @@
+        deg1 = poly(self.annihilator, self.generator).degree()
+        deg2 = poly(other.annihilator, other.generator).degree()
+        dim = max(deg1, deg2)
+        rowsself = [self.annihilator]; rowsother = [other.annihilator]
```
Multiple statements on a single line are not conventional

kshitij10496 commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,133 @@
+        for i, j in enumerate(sol):
+            solution += ((self.generator)**i)*j
+        return(HoloFunc((DifferentialOperator(solution)*DifferentialOperator(self.annihilator)), self.var))
+
```
Too long a line.
Use PEP008 style guide

kshitij10496 commented:
@shubhamtibra 
Nice work !
Although, do give this a read.
Its the preferred code styling guide [PEP 0008](https://www.python.org/dev/peps/pep-0008/)

shubhamtibra committed:
A preliminary representation of Holonomic Function and Differential Operator.

shubhamtibra committed:
Added Comments

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,161 @@
+        for i, j in enumerate(sol):
+            expr += ((self.gen)**i)*j
+
+        return(HoloFunc((DifferentialOperator(expr)*DifferentialOperator(self.ann)), self.var))
```
Don't put parentheses next to `return`. It should just be `return HoloFunc((DifferentialOperator(expr)*DifferentialOperator(self.ann)), self.var)`

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,161 @@
+        return(HoloFunc((DifferentialOperator(expr)*DifferentialOperator(self.ann)), self.var))
+
+        
+
```
The code quality tests should error from these extra lines at the end of the file.

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,161 @@
+            diff2 = (gen*DifferentialOperator(rowsother[-1])).eq
+            rowsother.append(diff2)
+            row = rowsself + rowsother
+            r = [[poly(expr, self.gen).nth(i) for i in xrange(dim + 1)] for expr in (row)]
```
Don't use `xrange`. Import `range` from `sympy.core.compatibility` and use that.

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,161 @@
+
+        # constructing the matrix of the ansatz
+        r = [[poly(expr, self.gen).nth(i) for i in xrange(dim + 1)] for expr in (row)]
+        from sympy.matrices import Matrix
```
Don't put imports in the middle of a function. Put them at the top of the file, or if there are circular import issues, at the top of the function.

asmeurer commented:
Your objects should subclass from `Expr` (or `Basic`, if they don't make sense as expressions), and follow the SymPy conventions. Look for other objects that subclass from Expr to see how this is done.

certik commented:
@shubhamtibra I think that this looks like a great start. Others have commented on technical issues.

What is the internal representation? A recursive polynomial in Dx? I wonder if a list of polynomials wouldn't be easier/more robust, i.e. a polynomial for each power of Dx. Then you can easily print it as needed and so on.

shubhamtibra commented:
Hi, @asmeurer .
I thought about subclassing from Expr, but is it  possible to create a  non commutative algebra this way?
@certik Inside the class, a holonomic function is stored as an SymPy expression with symbol `Dx` as the generator, initial conditions will be stored as a list and a seperate point for initial conditions can be provided(default is zero). Using a list of polynomials sounds better, I will be working on that.
One problem is that the Differential Operator can only be provided with all the `Dx` in right, any idea how to automatically convert a `Dx` at left to right using the commutation rule.

certik commented:
@shubhamtibra I don't understand what you mean by "on the right". You need to use the commutation rule to propagate all Dx the the left, e.g. Dx*x -> x*Dx.

shubhamtibra commented:
@certik 
In the class `DifferentialOperator`, a generator like `x*Dx + 1` can be provided where the `Dx` is in the right side of the coefficients only, but it won't work when we give a generator like `Dx*x + 1`( which is equal to `x*Dx +2`.
When an expression is passed to this class, is it possible to check all the generators which are at left and finally bringing them to the right using our commutation rule(or may be save it as it is).

Right now if `Dx*x +1` is given it will consider it the same as `x*Dx + 1`.

shubhamtibra committed:
Added functionality

Added a function to find list of coefficient polynomials from
recursive noncommutative expression and a function for list to expression.
Added functionality, now DifferentialOperator works even when generator is given in the
left side. Holonomic addition made more robust and solved few bugs.

shubhamtibra commented:
Now it also stores the list of polynomial coefficients of `Dx` and also solved the printing bug using this.

shubhamtibra commented:
The class `DifferentialOperator` works by creating a `Dx = symbols('Dx')` and then passing an expression in `Dx` and `x`.

    >>> from sympy import symbols
    >>> from sympy.functions.special.holonomic import DifferentialOperator
    >>> Dx, x = symbols('Dx, x')
    >>> DifferentialOperator(x)*DifferentialOperator(Dx+1)
    x + (x)*Dx**1

I would like to make it work where the user don't have to create a symbol and just using `Dx` in an expression will automatically create an instance of `DifferentialOperator`. e.g.

    >>> from sympy.functions.special.holonomic import DifferentialOperator
    >>> x = symbols('x')
    >>> Dx*x
    x*Dx + 1
    >>> isinstance( Dx*x, DifferentialOperator)
    True

Any idea how this can be done?
@certik What are your views on this kind of API?

shubhamtibra commented:
The class `DifferentialOperator` subclasses SymPy's `Expr`, and now also works when generator is in the left side, function `shifttoright` propagates all Dx to the right side using the commutation rule, as specified by Ondrej in the previous comments.
Solved few bugs for holonomic addition.
Please review and give your suggestions.
Ping @certik

shubhamtibra committed:
Added Normalization and changed printing

shubhamtibra commented:
I have added normalization in addition and soon I will be adding tests.
Should I add the DifferentialOperator class somewhere else or is it fine here?
Any suggestions and reviews would be truly helpful.
Ping @certik

shubhamtibra committed:
Fix build errors

shubhamtibra committed:
The variable `x` now is generator of the polynomial ring.

shubhamtibra committed:
Fixing Build Errors

shubhamtibra committed:
Added Docs

shubhamtibra commented:
Two tests are failing.

1. The arguments of the class `DifferentialOperator` are not instances of `Basic` so test_args is failing.
2. I added the decorator `@call_highest_priority` in the class `DMP` which is causing one test to fail.

ping @certik @jksuom .

jksuom commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,429 @@
+    def __init__(self, base_ring, generator):
+
+        self.base_ring = base_ring
+        self.diff_operator = DifferentialOperator([0, 1], self, generator)
```
0 and 1 could also be ``base_ring.zero`` and ``base_ring.one`` which have the correct type.

jksuom commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,429 @@
+                if isinstance(j, int):
+                    list_of_poly[i] = (
+                        (self.parent_ring).base_ring).from_sympy(S(j))
+                elif not isinstance(j, DMP):
```
Instead of ``DMP`` we could have ``parent_ring.base_ring.dtype``. That would be more flexible. (BTW, maybe it could be possible to drop the ``_ring`` parts for brevity, so just ``parent.base.dtype``.)

jksuom commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,429 @@
+        """
+
+        ring_0 = self.parent_ring.base_ring.from_sympy(S(0))
+        ring_1 = self.parent_ring.base_ring.from_sympy(S(1))
```
Here ``...base_ring.zero/one`` could be used.

shubhamtibra committed:
Further Corrections

jksuom commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,428 @@
+            list_self = self.listofpoly
+            list_other = other.listofpoly
+
+            if min(len(list_self), len(list_other)) is len(list_self):
```
This could be ``if len(list_self) <= len(list_other):``

And the two constructions could follow immediately. There is no need to set the ``minimum`` flag.

jksuom commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,428 @@
+    def __init__(self, list_of_poly, parent, generator=None, var=None):
+
+        if generator is None:
+            self.gen_symbol = symbols('Dx', commutative=False)
```
It seems to me that the generator symbol would more properly belong to the parent algebra. All differential operators in the same operator algebra should share the same generator.

shubhamtibra committed:
Holonomic Function addition integrated with Operators

shubhamtibra committed:
Generator is attribute of Algebra instead of the operator.

shubhamtibra committed:
Added Integration and Comments

shubhamtibra committed:
Added Tests

shubhamtibra commented:
Some tests are failing due to some error in printing. I am looking into them. And one test failed in `test_relational`.

shubhamtibra committed:
Fix build errors

shubhamtibra committed:
Remove Newline for Code Quality

shubhamtibra committed:
Remove the underscores

shubhamtibra commented:
Should we think about merging up to this ( Differential Operator and addition, integration of holonomic function) or should I proceed to the next step. If needed, I can work more on this and get the merged up to this point.
Ping @jksuom @certik .

jksuom commented:
I think the multiplication of differential operators should also be included before merging.

shubhamtibra commented:
That seems fine. I will start working on that.

shubhamtibra committed:
Added Multiplication with tests

shubhamtibra commented:
I haved added multiplication along with tests.

shubhamtibra commented:
@jksuom You meant multiplication of Holonomic Functions right? Differential Operators multiplication is already there.

shubhamtibra commented:
The class `DifferentialOperator` supports all addition, multiplication, subtraction and power. Holonomic function can be added and multiplied after last commit.

jksuom commented:
It seems that I misunderstood the comment `# if other is Dx` in `DifferentialOperator.__mul__` as implying that an `else` would have to be added for the more general case. After looking more closely I see that it is not needed. So, the operations for `DifferentialOperator` are supported.

shubhamtibra commented:
So I guess it should be reviewed so that we can think about merging.

shubhamtibra commented:
One problem is that the `args` of `DifferentialOperator` class are not instances of `Basic`( its a list and a ring) which is against SymPy's convention if a class subclasses from `Expr`. So to pass the test, I put an `@XFAIL` in `test_args.py`.

certik commented:
What is the motivation for the modifications in `sympy/polys/polyclasses.py`? Does it have some performance implications?

shubhamtibra commented:
It was to set priority for multiplication in `*` when one operand is `DifferentialOperator` and the other one is `DMP` type.

shubhamtibra commented:
So to make sure that the method in `DifferentialOperator` is called.

certik commented:
Do we have tests that test this?

shubhamtibra commented:
Yes. The code uses this priority mechanism for addition and multiplication. The function  in `test_DifferentialOperator()` at sympy/functions/special/tests/test_holonomic.py have tests for addition and multiplication.

jksuom commented:
I'm not sure that it would be necessary to add `_op_priority` to `DMP`. It is used when reading input expressions where the polynomials are ordinary SymPy expressions that have `_op_priority` set by default. 
What would happen if no changes were made in `polyclasses.py`?

certik commented:
@shubhamtibra in the first iteration of holonomic functions, I would prefer not to modify polynomials. Why don't we just do things internally somehow, like @jksuom suggested. Later in the summer, once we have a good solid module, then it would be time to think how to best integrate it.

shubhamtibra commented:
@jksuom I agree with you. I will be working on that.

shubhamtibra committed:
Revert changes in polyclasses.py

shubhamtibra commented:
I have made the changes.

shubhamtibra commented:
I have reverted the changes made in `sympy/polys/polyclasses.py`. This can be reviewed now. Ping @certik @jksuom .

jksuom commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,523 @@
+
+    _op_priority = 30
+
+    def __init__(self, list_of_poly, parent, generator=None, variable=None):
```
Is `generator` or `variable` needed somewhere?

jksuom commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,523 @@
+
+        if isinstance(other, DifferentialOperator):
+            # if other is `Dx`
+            if other.listofpoly == other.parent.derivative_operator.listofpoly:
```
Is is necessary to treat the `other == Dx` case separately?

shubhamtibra commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,523 @@
+        # using the commutation rule gen*b = b*gen + b'
+
+        def _diff_n_times(b):
+            expr = b * gen + b.diff()
```
Here we use the commutation rule `Dx*a = a*Dx + a'`. So in order to multiply two differential operators we need to find `a*Dx` (here other is `Dx`) which is implemented separately.

jksuom replied:
I believe it would be possible to compute the polynomial list of the product without forming intermediary `DifferentialOperator` objects like `a*Dx` or `b * gen`. Once the list is finished, the final product differential operator can be returned. It should not be necessary to transform the polynomials to expressions and back. Doing everything by means of the polynomial operations would be more efficient.

shubhamtibra replied:
Can you set an example for computing this without using expressions?

shubhamtibra replied:
Okay, I found a method where we don't have convert the list to expressions but it will still contain the special case of `other == Dx`. Is something like this would be efficient.

shubhamtibra replied:
Since the special case `other == Dx` works using polynomials without converting to expressions. Using `a*Dx` shouldn't affect the performance right?

jksuom replied:
I think it should not be necessary to deal with `Dx` while computing the polynomial list. On can loop over the list of `self` with `i in range(self.order + 1)`. For each `i` one should compute the list of `Dx^i * other` working with polynomials and their derivatives (without using objects of `DifferentialOperator` class). The resulting list is the multiplied elementwise with `self.listofpoly[i]`. All these partial lists are added to form the final list.

jksuom replied:
So the problem is the following: Given the polynomial lists of `self` and `other` compute the polynomial list of `self * other` using only ring operations and derivatives of polynomials.

shubhamtibra replied:
Are you seeing how will compute the list of `Dx^i * other`? Then it's done.

shubhamtibra committed:
Removed generator and variable from DifferentialOperator class.

shubhamtibra commented:
The `generator` and `variable

shubhamtibra committed:
Using only polynomials in DifferentialOperator.__mul__

shubhamtibra commented:
@jksuom Let me know how this looks to you now?

jksuom commented:
I added some line notes above, but maybe they were hidden.

shubhamtibra commented:
@jksuom Looks like both the ways (using DifferentialOperator objects and not using them) will are same in efficiency. For instance if one doesn't want to use DifferentialOperator objects then to compute list of Dx^i * other we will have to iterate through the list to differentiate each element then add to the partial list formed by shifting each element of the list to right by one. And this is exactly what we are doing now.

shubhamtibra commented:
Here the case `other == Dx` just shifts each element of the list to right by one. If we don't want to write this case explicitly we will have to do the shifting anyway while computing `Dx^i * other`. So I guess it's the same thing.

jksuom commented:
I guess that the polynomial manipulations must be more or less the same. But I think that not constructing intermediary differential operators and just working with the lists could be more efficient.

It is true that `other == Dx` is not an expensive operation, but I'm not sure that it should be treated separately. In the computation of `Dx^i * other` the polynomials (or their derivatives) become 'shifted' in any case.

shubhamtibra committed:
Working with lists

shubhamtibra commented:
@jksuom I have made the changes. Let me know your views.

jksuom commented:
Thanks, that was quick! It looks promising to me. Maybe `_add_lists` could be moved out and used also for `__add__`?

shubhamtibra committed:
Making _add_lists() global

shubhamtibra commented:
So, the tests have passed. What should be our next move?

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,524 @@
+
+class DifferentialOperator(Expr):
+    """
+    The operator can be created by providing a list of polynomials for each power of Dx
```
Please make the first line of the docstring describe what the class is.

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,524 @@
+                powreduce = self**(n / 2)
+                return powreduce * powreduce
+
+    def str(self):
```
Please use the normal printing mechanisms (define the printer in `sympy/printing/str.py`).

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,524 @@
+
+    repr = str
+
+    def __eq__(self, other):
```
99% of the time it is a bad idea to redefine `__eq__` on a `Basic` subclass. Can you explain what this does and justfiy why it's needed?

asmeurer commented on sympy/core/tests/test_args.py:
```diff
@@ -1710,6 +1710,14 @@ def test_sympy__functions__special__hyper__HyperRep_sinasin():
+@XFAIL
+def test_sympy__functions__special__holonomic__DifferentialOperator():
+    from sympy.functions.special.holonomic import DifferentialOperator, DiffOperatorAlgebra
+    from sympy.polys.domains import PythonIntegerRing, ZZ
```
PythonIntegerRing is not used here.

asmeurer commented on sympy/core/tests/test_args.py:
```diff
@@ -1710,6 +1710,14 @@ def test_sympy__functions__special__hyper__HyperRep_sinasin():
     assert _test_args(HyperRep_sinasin(x, y))
 
 
+@XFAIL
```
This is not good. We shouldn't be creating new Basic subclasses that don't satisfy the args invariants.

asmeurer commented on sympy/printing/str.py:
```diff
@@ -728,6 +728,10 @@ def _print_Tr(self, expr):
         return "%s(%s)" % ("Tr", self._print(expr.args[0]))
 
+    def _print_DifferentialOperator(self, expr):
+        print_str = expr.str()
```
Put the actual printer here. Also, the `str` printer should always be valid Python, which recreates the class (use the pretty printer for fancy printing). And make sure to use `self._print` to recursively print subexpresions.

asmeurer commented:
This looks like a good start, but we need to think more carefully about the design of the DifferentialOperator class, especially if it's going to be a Basic subclass.

asmeurer commented on sympy/functions/special/holonomic.py:
```diff
@@ -0,0 +1,524 @@
+    return (ring, ring.derivative_operator)
+
+
+class DifferentialOperatorAlgebra:
```
Every class should subclass from `object`. Otherwise in Python 2 it will be an old-style class.

asmeurer commented:
I think there are two directions we can go here:

- Make the differential operator module like the polys, i.e., distinct from the core. That means nothing subclasses from Basic/Expr. The upside here is that it is easy to include the polys objects directly, because they are also separate from the core. The downside is that it won't be possible to combine holonomic function objects with other SymPy objects. If we do this, the module should be its own sympy submodule (like sympy.holonomic, instead of sympy.functions.special.holonomic). 

- Make everything compatible with the core. This means that everything subclasses from Expr and follows the args invariants (all args are Basic subclasses, and all objects are rebuildable from their args). The advantage here is that it will be possible to mix holonomic objects with other expressions. The downside is that anything that uses the polys will have to be done indirectly, since polys objects aren't Basic, so they can't be part of .args. 

I would prefer the second way, if we can work out how to do it.  That way we can actually represent holonomic functions as functions, and use them in expressions. Even if we don't do it that way, we would likely want a wrapper that does work like that, so it's just a question if it's useful to have a layer that works without Basic subclassing. 

Either way, you should go one way or the other. Classes that subclass from Expr but don't really behave like Basic subclasses should are a recipe for disaster.

shubhamtibra commented:
Thank you @asmeurer for you reviews. I will work on your suggestions. Regarding the two ways we can go, I agree that the latter one is better i.e. keeping it integrated with core. Also @certik suggested first to go as a separate module and then integrating it with SymPy after it is mature.

One option is to integrate it with core now or we will do it later anyway.

shubhamtibra commented:
@asmeurer If the class doesn't subclasses from `Basic` or `Expr` then should we put it's printer in the class as `__str__` itself?

shubhamtibra committed:
Now the class DifferentialOperator doesn't subclasses from `Expr`.

asmeurer commented:
If you don't use Basic subclassing, then you can use your own printing methods. 

I would suggest making everything separate from the core to start with (no Basic or Expr subclassing), and once things have stabilized, create an Expr wrapper class. This way you can just use the polys directly, and you won't need to worry about most design issues. The downside is that until you have a wrapper, it will be difficult for people to play with this code. 

So I would suggest: 

- Remove the Expr subclass from DifferentialOperator.

- Move the module to `sympy.holonomic` (or `sympy.polys.holonomic`). 

- Use exclusively the polys for internal expressions. Holonomic functions only require polynomials to represent, right?

shubhamtibra commented:
Yes. The internal representation of Holonomic function uses a `DifferentialOperator` object which eventually uses polynomials for it's internal representation.

shubhamtibra committed:
Making Holonomic a seperate module

shubhamtibra commented:
Thanks @asmeurer . I did what you suggested. Can you give me an example of such type of wrapper if we have in SymPy so that I can take a look on it.

certik commented:
Thanks @shubhamtibra, now this PR looks nice, everything is local. As @asmeurer said, once this is more mature, we can think about how to best integrate it with the rest of SymPy.

asmeurer commented:
The only example I can think of off the top of my head is `RootOf`.

jksuom commented:
Also the functions in the beginning of `rings.py` and `fields.py` are of the wrapper type.

shubhamtibra commented:
The tests have passed. What should be our next move? Ping @jksuom @certik .

jksuom commented:
The next step could be making `HoloFunc` work with polynomials (and their fractions) instead of SymPy expressions.

shubhamtibra commented:
Do we have solvers that can solve linear systems in polynomials instead of expressions?

certik commented:
First of all, great job! It looks very good overall.

Some minor comments first:

I noticed you use 8 spaces at one point. Use 4.

Put space after comma in tests, e.g. change ",x" to ", x".

The docstrings should be improved. Add doctests as documentation. Lookup
documentation how to write docstrings, e.g. a blank line after the first
line, etc.

Why not to call the class HolonomicFunction? It's a bit longer, but more
descriptive I think.

What next:

I would first implement the initial conditions in HoloFunc and make sure
they work for addition and multiplication.

How robust is the addition and multiplication?

Then add a conversion from a hypergeometric function. See the wiki I wrote
for code that does it.

I would worry about using polynomials after the above is implemented. Since
the above should be enough to start implementing some new algorithms, like
integration.

Sent from my mobile phone.
On May 12, 2016 12:41 AM, "Subham Tibra" <notifications@github.com> wrote:

> Do we have solvers that can solve linear systems in polynomials instead of
> expressions?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/10793#issuecomment-218672964>
>

shubhamtibra commented:
Thanks @certik . I will work on your suggestions.
Regarding robustness, the algorithm performs slow when the power of `Dx` and `x`  are increased due the increased complexity of linear system. If we can solve the linear system quickly then that can improve a lot in performance.

The Initial Conditions can be implemented for the case when both the points at which the values are known are same. I'll be working on that. Regarding the cases when the points are not same, we might need symbolic and numerical values of the function which isn't implemented yet.

shubhamtibra committed:
Improved Docstrings and HoloFunc to HolonomicFunction

shubhamtibra commented:
What should be the input API for Initial Conditions. Is a string based API good like `"y(0) = 1"` or a dictionary like `{ y(0) : 1 }`.

certik commented:
I would do x0=0, y0=1 for the point and value. Lookup scipy initial
conditions for ode integration. Let's follow the same conventions. Also
look into sympy's ode modules. Don't use strings.

Sent from my mobile phone.
On May 13, 2016 6:14 AM, "Subham Tibra" <notifications@github.com> wrote:

> What should be the input API for Initial Conditions. Is a string based API
> good like "y(0) = 1" or a dictionary like { y(0) : 1 }.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/10793#issuecomment-219026816>
>

shubhamtibra committed:
Added Initial Conditions for Addition with Tests

shubhamtibra commented:
I have added Initial Conditions for addition and will soon be adding for multiplication. Ping @certik @jksuom .

jksuom commented on sympy/holonomic/holonomic.py:
```diff
@@ -577,6 +644,39 @@ def __mul__(self, other):
+
+def _extend_y0(Holonomic, n):
+    """
+    Tries to find more initial conditions by substituting the inital
```
inital -> initial

jksuom commented:
Looks good. Only, the conditions lists would look better when separated with commas, I think.

shubhamtibra commented:
@jksuom Did you mean adding commas in printing between `f(0) = 2` and `f'(0) = 1`.

jksuom commented:
Yes, and also before `f(0) = 1`:
`HolonomicFunction((-1) + (1)Dx + (-1)Dx**2 + (1)Dx**3, x), f(0) = 1, f'(0) = 2`.

shubhamtibra committed:
Added Initial Conditions for Multiplication

shubhamtibra commented:
I have also added the support of Initial Conditions for multiplication. All test are tested from Maple's `gfun` package and `ore_algebra` in Sage.

shubhamtibra commented:
Where should I put the code for converting Hypergeometric Function. Should it be an instance method in `sympy/functions/special/hyper.py` or a function in `sympy/holonomic/holonomic.py`.

certik commented:
For now, just make it a function in a holonomic file.

Sent from my mobile phone.
On May 14, 2016 10:12 AM, "Subham Tibra" <notifications@github.com> wrote:

> Where should I put the code for converting Hypergeometric Function. Should
> it be an instance method in sympy/functions/special/hyper.py or a
> function in sympy/holonomic/holonomic.py.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/10793#issuecomment-219228663>
>

certik commented on sympy/holonomic/tests/test_holonomic.py:
```diff
@@ -0,0 +1,122 @@
+    q = HolonomicFunction(Dx**2 + 1, x, 0, [0, 1])
+    r = HolonomicFunction(2 -2*Dx + Dx**2, x, 0, [0, 2])
+    assert p * q == r
+    q = HolonomicFunction(x*Dx**2+1+2*Dx,x,0,[0,1])
```
space after comma

certik replied:
@shubhamtibra I merged it, but can you please still address this comment in a new PR?

certik commented:
@shubhamtibra, now it looks like a good time to merge this. Any objections?

Thanks for implementing the initial conditions.

shubhamtibra commented:
I closely monitored the time taken by code to execute. Turns out the operation `_normailze` is sucking a lot of time and the second most time consumer is solving the linear system i.e. `gauss_jordan_solve`. I will try to optimize that and then we can merge it. Any suggestions on how to do it?

shubhamtibra commented:
The solution of the linear system is a list of extremely complex expressions containing rationals and high powers of `x`. Is there a way we can normalize these functions to polynomials efficiently.

jksuom commented:
I believe the solution would be not to work with expressions in internal operations but solely with polynomials and their fractions. The corresponding change in ring series (by Shivam Vats) cut the time to a fraction.

shubhamtibra commented:
I think you are right. I have a question, How an expression like `x + 2 / (x + 2)` will be represented using `polys`.

jksuom commented:
There is a fraction field associated with each polynomial ring. Its elements are represented by pairs of polynomials: numerator and denominator. As it is a field, all the four rational operations, addition, subtraction, multiplication and division are implemented, and so e.g. `gauss_jordan_solve` can use its elements as such.
The fraction field of a `GlobalPolynomialRing` `R` (such as `QQ.old_poly_ring`) can be created by `K = R.get_field()`. A polynomial `p` of `R` can be directly converted to an element of `K` by `K.new(p.rep)`. If `p` and `q` are two polynomials of `R`, the fraction `p/q` can be obtained by `K.new((p.rep, q.rep))`. (There exist also `K.from_sympy` and `K.to_sympy`.)

shubhamtibra commented:
Thanks @jksuom , that was helpful. I am looking into this, so all the internal operations just uses Polynomials `DMP` and the Fractions and `DMF`.

certik commented:
Before you move to polynomials, can we merge this PR?

Sent from my mobile phone.
On May 15, 2016 6:38 AM, "Subham Tibra" <notifications@github.com> wrote:

> Thanks @jksuom <https://github.com/jksuom> , that was helpful. I am
> looking into this, so all the internal operations just uses Polynomials
> DMP and the Fractions and DMF.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/10793#issuecomment-219283488>
>

shubhamtibra commented:
Yes we can. +1

certik commented:
Ok, It's in (but see my comment above). This is a good start. Now we can move to use polynomials, as that should speed things up considerably.

shubhamtibra commented:
Thank you so much @certik . I will put the spaces in next commit. Should a new PR be opened for further addition of code?

certik commented:
Yes, open a new PR for every isolated feature. Do not combine multiple
features in the same PR.

Sent from my mobile phone.
On May 15, 2016 7:38 AM, "Subham Tibra" <notifications@github.com> wrote:

> Thank you so much @certik <https://github.com/certik> . I will put the
> spaces in next commit. Should a new PR be opened for further addition of
> code?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/10793#issuecomment-219286309>
>
